index.html

```diff
@@ -0,0 +1,552 @@
+          </dl>
+          <section data-dfn-for="SmartCardReaderStateFlagsIn">
+            <h5><dfn>SmartCardReaderStateFlagsIn</dfn> dictionary</h4>
+
+            <pre class="idl">
+              dictionary SmartCardReaderStateFlagsIn {
+                boolean unaware = false;
+                boolean ignore = false;
+                boolean unavailable = false;
+                boolean empty = false;
+                boolean present = false;
+                boolean exclusive = false;
+                boolean inuse = false;
+                boolean mute = false;
+              };
+            </pre>
+
+            <dl>
+              <dt><dfn>unaware</dfn> member</dt>
+              <dd>The application is unaware of the current state, and would like
+                to know.</dd>
+
+              <dt><dfn>ignore</dfn> member</dt>
+              <dd>The application is not interested in this reader, and it should
+                not be considered during monitoring operations.</dd>
+
+              <dt><dfn>unavailable</dfn> member</dt>
+              <dd>The application believes that this reader is not available for
+                use.</dd>
+
+              <dt><dfn>empty</dfn> member</dt>
+              <dd>The application believes that there is not a card in the
+                reader.</dd>
+
+              <dt><dfn>present</dfn> member</dt>
+              <dd>The application believes that there is a card in the
+                reader.</dd>
+
+              <dt><dfn>exclusive</dfn> member</dt>
+              <dd>The application believes that the card in the reader is
+                allocated for exclusive use by another application.</dd>
+
+              <dt><dfn>inuse</dfn> member</dt>
+              <dd>The application believes that the card in the reader is in use
+                by one or more other applications, but may be connected to in
+                shared mode. </dd>
+
+              <dt><dfn>mute</dfn> member</dt>
+              <dd>The application believes that there is an unresponsive card in
+                the reader.</dd>
+            </dl>
+          </section>
+        </section>
+        <section data-dfn-for="SmartCardReaderStateOut">
+          <h4><dfn>SmartCardReaderStateOut</dfn> dictionary</h4>
+          <p>The actual state of a smart card reader.</p>
+
+          <pre class="idl">
+            dictionary SmartCardReaderStateOut {
+              required DOMString readerName;
+              required SmartCardReaderStateFlagsOut state;
+              required ArrayBuffer answerToReset;
+            };
+          </pre>
+
+          <dl>
+            <dt><dfn>readerName</dfn> member</dt>
+            <dd>Name of the smart card reader.</dd>
+
+            <dt><dfn>state</dfn> member</dt>
+            <dd>The actual state of that smart card reader.</dd>
+
+            <dt><dfn>answerToReset</dfn> member</dt>
+            <dd>The inserted card's [[ISO7186-3]] Answer To Reset (ATR), if any.</dd>
```

Comment:
That's something I was wondering about actually. Right now my idea is that if there's no ATR, that `ArrayBuffer` will be empty (zero bytes). So one can just check for the returned buffer size and that's all.

But if this is made optional, we would technically have two different ways of expressing the same thing: an absent ATR could be either a `null` `anserToReset` or an empty one. Not sure if checking for null is more convenient than checking for emptyness. Also, if we go the optional/null way, we must say in the spec that a non-null ATR will never be empty, right? So that application code can safely make that assumption.

In the end I guess it's a matter of convention. And I'm not sure which way Web APIs go in that case. What do you think?

Reply:
This is also the convention I'm using on the SmartCardConnection.control() method (empty buffer).

Reply:
I would differentiate between states where an ATR cannot be returned (e.g. card or reader not present) and states where it can be returned but may be empty. How is this handled in the PC/SC API? Is the ATR something you have to query separately (and so requesting it can fail, i.e. no ATR present) or is it a static buffer that is always available but might be empty?

Reply:
As soon as the reader powers up (or resets) a card, the card is expected to automatically send back its identification bytes (whose format is mostly standardized). Answering to a reset is an integral part of a ISO compliant smart card. Not doing so is an unrecoverable error and communication with such an unresponsive card is not possible. The cause could be a damaged card or the reader applying a class operating conditions (voltage) not supported by it.

My understanding is that in PC/SC, a card that doesn't answer to a reset will cause the relevant API calls to fail with `SCARD_W_UNRESPONSIVE_CARD`: "The card is not responding to a reset."

Reply:
> I would differentiate between states where an ATR cannot be returned (e.g. card or reader not present) and states where it can be returned but may be empty.

And such an unresponsive card would yield the `"present" AND "mute"` state, where `answerToReset` cannot have a value (would be inconsistent).

In summary, I see two classes of states:
* states where ATR must not be empty (eg: card present and in use)
* states where an ATR does not apply (eg: no card present, card present but mute)

Reply:
If the card may never return an empty ATR that seems even better to me. We can say that `answerToReset` is either present and non-empty or irrelevant for the current reader state (and therefore `undefined`). That way we never need to construct an artificial empty ATR value when that isn't something actually allowed by the specs.

Reply:
Ok.